airbyte-server/src/main/java/io/airbyte/server/handlers/DestinationDefinitionsHandler.java

```diff
@@ -141,8 +144,42 @@ public DestinationDefinitionRead getDestinationDefinition(final DestinationDefin
         configRepository.getStandardDestinationDefinition(destinationDefinitionIdRequestBody.getDestinationDefinitionId()));
   }
 
-  public DestinationDefinitionRead createCustomDestinationDefinition(final DestinationDefinitionCreate destinationDefCreate)
+  public DestinationDefinitionRead getDestinationDefinitionForWorkspace(
+                                                                        final DestinationDefinitionIdWithWorkspaceId destinationDefinitionIdWithWorkspaceId)
+      throws ConfigNotFoundException, IOException, JsonValidationException {
+    final UUID definitionId = destinationDefinitionIdWithWorkspaceId.getDestinationDefinitionId();
+    final UUID workspaceId = destinationDefinitionIdWithWorkspaceId.getWorkspaceId();
+    if (!configRepository.workspaceCanUseDefinition(definitionId, workspaceId)) {
+      throw new IdNotFoundKnownException("Cannot find the requested definition with given id for this workspace", definitionId.toString());
+    }
+    return getDestinationDefinition(new DestinationDefinitionIdRequestBody().destinationDefinitionId(definitionId));
+  }
+
+  public DestinationDefinitionRead createPrivateDestinationDefinition(final DestinationDefinitionCreate destinationDefCreate)
       throws JsonValidationException, IOException {
+    final StandardDestinationDefinition destinationDefinition = destinationDefinitionFromCreate(destinationDefCreate)
+        .withPublic(false)
+        .withCustom(false);
+    configRepository.writeStandardDestinationDefinition(destinationDefinition);
+
+    return buildDestinationDefinitionRead(destinationDefinition);
+  }
+
+  public DestinationDefinitionRead createCustomDestinationDefinition(final CustomDestinationDefinitionCreate customDestinationDefinitionCreate)
+      throws JsonValidationException, IOException {
+    final StandardDestinationDefinition destinationDefinition = destinationDefinitionFromCreate(
+        customDestinationDefinitionCreate.getDestinationDefinition())
+            .withPublic(false)
+            .withCustom(true);
+    configRepository.writeStandardDestinationDefinition(destinationDefinition);
+    configRepository.writeActorDefinitionWorkspaceGrant(
```

Comment:
I think we should consider writing a `configRepository` method that can perform these two steps in a single transaction. This seems like the kind of operation we want to either fully succeed or fully roll back

Reply:
yeah that makes sense, will add